client/src/main/scala/com/aliyun/emr/rss/client/write/LifecycleManager.scala

```diff
@@ -992,65 +1047,65 @@ class LifecycleManager(appId: String, val conf: RssConf) extends RpcEndpoint wit
       candidates: List[WorkerInfo]): WorkerResource = {
-    if (candidates.size < 1 || (ShouldReplicate && candidates.size < 2)) {
-      logError("Not enough candidates for revive")
-      return null
-    }
     val slots = new WorkerResource()
     allocateFromCandidates(reduceId, oldEpochId, candidates, slots)
     slots
   }
 
-  def reallocateSlotsFromCandidates(
+  private def reallocateSlotsFromCandidates(
       oldPartitions: List[PartitionLocation],
       candidates: List[WorkerInfo]): WorkerResource = {
-    if (candidates.size < 1 || (ShouldReplicate && candidates.size < 2)) {
-      logError("Not enough candidates for revive")
-      return null
-    }
-
     val slots = new WorkerResource()
     oldPartitions.foreach { partition =>
       allocateFromCandidates(partition.getReduceId, partition.getEpoch, candidates, slots)
     }
     slots
   }
 
-  def destroyBuffersWithRetry(
+  /**
+   * For the slots that need to be destroyed, LifecycleManager will ask the corresponding worker
+   * to destroy related FileWriter.
+   *
+   * @param applicationId application id
+   * @param shuffleId shuffle id
+   * @param slotsToDestroy worker resource to be destroyed
+   * @return destroy failed master and slave location unique id
+   */
+  private def destroyBuffersWithRetry(
       applicationId: String,
       shuffleId: Int,
-      worker: WorkerResource): (util.List[String], util.List[String]) = {
-    val failedMasters = new util.LinkedList[String]()
-    val failedSlaves = new util.LinkedList[String]()
+      slotsToDestroy: WorkerResource): (util.List[String], util.List[String]) = {
+    val failedMasterLocationUniqueIds = new util.LinkedList[String]()
+    val failedSlaveLocationUniqueIds = new util.LinkedList[String]()
 
     val shuffleKey = Utils.makeShuffleKey(applicationId, shuffleId)
-    worker.asScala.foreach { case (workerInfo, (masterLocation, slaveLocation)) =>
+    slotsToDestroy.asScala.foreach { case (workerInfo, (masterLocations, slaveLocations)) =>
       val destroy = Destroy(shuffleKey,
-        masterLocation.asScala.map(_.getUniqueId).asJava,
-        slaveLocation.asScala.map(_.getUniqueId).asJava)
+        masterLocations.asScala.map(_.getUniqueId).asJava,
+        slaveLocations.asScala.map(_.getUniqueId).asJava)
       var res = requestDestroy(workerInfo.endpoint, destroy)
       if (res.status != StatusCode.Success) {
         logDebug(s"Request $destroy return ${res.status} for " +
           s"${Utils.makeShuffleKey(applicationId, shuffleId)}")
         res = requestDestroy(workerInfo.endpoint,
           Destroy(shuffleKey, res.failedMasters, res.failedSlaves))
       }
-      failedMasters.addAll(res.failedMasters)
-      failedSlaves.addAll(res.failedSlaves)
+      failedMasterLocationUniqueIds.addAll(res.failedMasters)
+      failedSlaveLocationUniqueIds.addAll(res.failedSlaves)
     }
-    (failedMasters, failedSlaves)
+    (failedMasterLocationUniqueIds, failedSlaveLocationUniqueIds)
```

Comment:
We ignored the result of the destroyed message. So I think we can remove the result list.